### DISCUSSION TITLE
When does the Controller value update?

### DISCUSSION THREAD

**Discussion by alessandrojcm** [react-hook-form/react-hook-form#2787] [Q&A]
Hi there!

I'm having this weird issue with the `Controller` API, I'm not sure if it is a bug or if I didn't get something. I have the following input:
![image](https://user-images.githubusercontent.com/6154350/91789696-2a684e00-ebdd-11ea-83fa-646a8aa7bce5.png)

They are both native HTML `input` tags, and I have them like so (some tags removed for clarity):
```jsx
<Controller
    name={'startDate'}
    control={control}
    render={({ onChange, value }) => (
        <input
            type="date"
            // format from date-fns
            value={format(value, 'yyyy-MM-dd')}
            onChange={(t) => {
                // Some validations here

                // Using date-fns with the functional programming API to take the Date from the input, and generate a new date but
                // without overriding the time 
                onChange(
                     set(parseISO(t.target.value), {
                          minutes: value.getMinutes(),
                          hours: value.getHours(),
                      })
                );
                // There's another input that whose validation depends on this one, so I trigger it right away
                trigger('endDate');
            }}
        />
    )}
/>
<Controller
    name={'startDate'}
    control={control}
    render={({ onChange, value }) => (
        <input
            type="time"
            value={format(value, 'hh:mm')}
            onChange={(e) => {
                const [hour, minute] = e.target.value.split(':');

                // Setting the time to the date that's already on the form
                onChange(
                    set(value, {
                        minutes: Number(minute),
                        hours: Number(hour),
                    })
                );
                trigger('endDate');
            }}
        />
    )}
/>
```

So when the user inputs a date from the left side input, what I do is create a new date with the date from the input and the time from the `value`; so the time from the right side doesn't get overridden. For the right input, I do the opposite: I take the time from the input, the current date from the form, and generate a new date with the new time.

Now, my problem is the following: when I try to input the time, the date input resets to the previous date. So, for example, the date is 27th of august and I change it to the 28th and then I go to change the time the date gets set to the 27th again; it's like the `value` passed to the render method of `Controller` is not up to date. On the contrary, if I do `watch('startDate')` and use that instead if works properly; so it's something with `Controller`. Am I missing something here?

I'm using Yup and the validation method is `onChange`, in case that's relevant. Thanks!

**Comment by bluebill1049** [comment]
why are you running `tiger` inside the `onChange`? shouldn't mode: 'onChange' take care of that?

**Reply by alessandrojcm** [reply]
Hi! Yes, normally `onChange` would take care of that. But I observed that it only triggers validation for the field I just changed (`startDate`, in this case). I have another field just like this (called `endDate`) whose validation depends on the value of `startDate`; so as soon as `startDate` changes I manually trigger the validation of `endDate` in order for the validations messages to appear. Otherwise, they only appear when you edit the `endDate` field.

**Reply by bluebill1049** [reply]
ok, make you reproduce this in a codesandbox (keep it simple plz) i will take a look the root cause.

**Reply by alessandrojcm** [reply]
There you go: https://codesandbox.io/s/blissful-wescoff-86wjr
Try changing the day of the "This doesn't work" section and then change the hour, the date will reset

**Reply by bluebill1049** [reply]
Thanks @alessandrojcm , why you need controller for native input?

**Reply by bluebill1049** [reply]
what's the steps to reproduce the issue?

**Reply by alessandrojcm** [reply]
> Thanks @alessandrojcm , why you need controller for native input?

Two reasons: 1) an input of type date/time returns a string, I need to parse it as a date to run validations 2) I need to trigger the validation for another field (`endDate`) manually; I can add that also to the demo if you need it

**Reply by alessandrojcm** [reply]
> what's the steps to reproduce the issue?

Go to the inputs below the "This doesn't work", change the day or the month on the left input and then try to change the hour on the right input; the date on the left will reset.

**Reply by bluebill1049** [reply]
Thanks, will take a look at it.

**Reply by alessandrojcm** [reply]
Hey! Any updates on this? @bluebill1049

**Reply by bluebill1049** [reply]
we have valueAsDate now, not sure if this will help your issue? https://codesandbox.io/s/vigorous-hellman-ju4j4?file=/src/App.js

**Reply by alessandrojcm** [reply]
I would need to check, I remember we locked the `react-hook-form` version due to some typing errors with yup; but I cannot recall what was the issue exactly.

**Reply by bluebill1049** [reply]
right, i think now that we support valueAsDate for native input like Date, there shouldn't be an issue to know when Controller(middle component) when it's update value.
